### PR TITLE
added find longest match method

### PR DESCRIPTION
**Pull request by jlpedrosa**
This is a method to get only the longest exact match.

### CODE REVIEW

**Comment by npgall** [comment]
Hi Joseluis,

This sounds interesting!

To merge this I'd need the following:
* Unit tests! This pull request does not have any. This library has 100% test coverage at the moment so pull requests need to maintain this level of coverage.
* Consistency with existing APIs. The methods in this PR are not consistent with the triplets of methods for finding keys, finding values, and findings keys-and-values. People might like to be able to look up the longest key, the longest value for the key, or both together. This PR only supports looking up the value.
* Example usage and documented behaviour. The JavaDocs could maybe provide example usage. At the moment I am a bit confused about the behaviour actually, because AFAICS the new method does not examine all of the characters inside each node - it just examines the last character? The behaviour would be clearer if you could provide unit tests or examples.

I really appreciate contributions via pull request, so many thanks! Let me know if you need more info or help.

Best regards,
Niall

**Comment by jlpedrosa** [comment]
Niall

about point 1 and 3) I can work on that after we agree on 2)

The idea is the same as https://en.wikipedia.org/wiki/Trie (using their example tree in the header)
if I look for "Tee" as input parameter,  it should return "te" as.

"tennessee" should return "ten". "Hello" which is not found, should return "". It keeps progressing in the tree while the tree has an option that matches the pattern, it does not look only for the latest one. 

Makes sense? 

JL

**Comment by npgall** [comment]
I see. It might be useful if we could explain the proposed behaviour with examples as was done in issue #2 and issue #5. Is this much different than the getKeysPrefixing() method discussed in issue #5?

The Trie data structure shown on wikipedia differs from the RadixTree though as the RadixTree contains more than one character in each node. So I think you will need to examine the characters inside each node.

If it helps the searchTree() method *might* be useful - it currently returns the number of characters matched.

**Comment by jlpedrosa** [comment]
Hi Nail,

The thing is, for the use case I need,  there will be a lot of keys that would match, I just need ONE and the longest, the tree will contain 50.000+ entries.  I can get may entries back from a query, (and filtering the longest would defeat the properties of the tree). It is for prefix matching of phones, maybe it simpler if you´d imagine as a router table (but based on strings instead of binary... ok the example is shit) 

I got worried when you said the node can have more than one character.... that was not obvious from the interface, I thought it was same as wikipedia, I'll improve the implementation and come back to you.

After fixing it, adding the documentation and reaching 100% of unit test, would you accept the PR?

Thanks!

**Comment by jlpedrosa** [comment]
Hi Niall

I've made a new commit, I've already added the unit tests and checked for the whole string. There is part of the method that could go to CharSequences util classes, but despite of it. Does it look good now?
If you'd like to avoid the ping pong we can continue over skype and add here the conclusion... or if you want I can create an Issue. 

Best 

JL

BTW, I just saw I miss spelled your name in the other posts, my apolgies.

**Comment by npgall** [comment]
Hi Joseluis,

The new commit looks good. And I can see that it works :)

A few points:
* The new code still just returns the value. Can you add methods to also return the key, the value, and both?
* Would you mind moving the new code into the ConcurrentInvertedRadixTree class? As these methods will be closely related to the existing getKeysPrefixing() method there. 
* How about we call the new methods: getLongestKeyPrefixing(), getValueForLongestKeyPrefixing(), getKeyValuePairForLongestKeyPrefixing?
* The new code uses variable names such as "tVal" and "idxMatch"; whereas the existing code in this library uses descriptive variable names like "charsMatched". I'd prefer if the new code followed this convention.
* As this feature does the same thing as calling ConcurrentInvertedRadixTree.getKeysPrefixing() except it returns only the last key, could the new code leverage the existing algorithm?
    * The existing algorithm can be found in [scanForKeysAtStartOfInput](https://github.com/npgall/concurrent-trees/blob/master/code/src/main/java/com/googlecode/concurrenttrees/radixinverted/ConcurrentInvertedRadixTree.java#L50)
    * It's not an exact fit though, because the purpose of your new methods is to _efficiently_ locate the last key. The existing code _will not be as efficient_ for this use case because it will allocate objects for the earlier keys found. But the underlying algorithm in both implementations will be the same. 
    * Would it be possible to adapt the code style in your implementation to use the same variable names and naming conventions as the existing one?

Basically consistency is the name of the game.

BTW those tests are exactly what I was looking for, and it explains the behaviour very well. Many thanks!

This will be a really worthwhile addition Joseluis - many thanks!

**Comment by jlpedrosa** [comment]
Hi Niall

> The new code still just returns the value. Can you add methods to also return the key, the value, and both?<

Done

> Would you mind moving the new code into the ConcurrentInvertedRadixTree class? As these methods will be closely related to the existing getKeysPrefixing() method there.<
 
I'm not sure if I agree with you in this point, let me explain. This prefix lookup is one of the reasons of the TRIE/Radix tree data structure was "created", so I think it has to be AT LEAST in radix tree, also as the interface claims,` interface InvertedRadixTree<O> extends RadixTree<O> `  so I think it really makes more sense to put it in RadixTree as it is the base of the next structures. 

> How about we call the new methods: getLongestKeyPrefixing(), getValueForLongestKeyPrefixing(), getKeyValuePairForLongestKeyPrefixing?<

Done.

>The new code uses variable names such as "tVal" and "idxMatch"; whereas the existing code in this library uses descriptive variable names like "charsMatched". I'd prefer if the new code followed this convention.<

Done

>As this feature does the same thing as calling ConcurrentInvertedRadixTree.getKeysPrefixing() except it returns only the last key, could the new code leverage the existing algorithm?
    The existing algorithm can be found in scanForKeysAtStartOfInput
    It's not an exact fit though, because the purpose of your new methods is to efficiently locate the last key. The existing code will not be as efficient for this use case because it will allocate objects for the earlier keys found. But the underlying algorithm in both implementations will be the same.
    Would it be possible to adapt the code style in your implementation to use the same variable names and naming conventions as the existing one?<

Agree, it's not the same and also having the stack bouncing through the iterator to get the latest is also not the best. I've tried to put it as similar to the others as I could. Please have a second look, if not, please feel free to rename them to proper English. 

All tests are passing, let me know what you think. 


BR 
JL

Side note: my name is Jose Luis (a lot of people in Spain have composite?/compund? names),

**Comment by npgall** [comment]
Hi Jose Luis,

The distinction between RadixTree and InvertedRadixTree is that, while both classes share the same data structure, they use different traversal algorithms. The RadixTree uses the `searchTree()` algorithm to scan the tree for keys which match the input, whereas the InvertedRadixTree uses the `scanForKeysAtStartOfInput()` algorithm to scan the input for keys which are contained in the tree. So I'd prefer to locate this new feature in InvertedRadixTree as it uses the same type of traversal algorithm.

There is no downside for users - people who want to use the same data structure with both algorithms can just use the InvertedRadixTree because it extends the RadixTree.

That commit looks good. One suggestion is, can you remove the `CharSequence remaining` object? You can see in the `scanForKeysAtStartOfInput()` method it uses `charsMatched` to keep track of the position in the input, instead of generating a new CharSequence for the remaining characters in each iteration. It will be beneficial to avoid creating objects during the traversal when possible, to reduce the overhead for GC.

Thanks for adding the triplet of methods. If you're happy to do the two minor changes above I'll be happy to go ahead and merge. And I appreciate this new feature!

Cheers,
Niall

**Comment by jlpedrosa** [comment]
Hi Niall

I'll be traveling for a couple of days, I think I'll be able to do those
changes by thursday. When do you think it could be included in a new
release?

BR

JL

**Comment by npgall** [comment]
Hi Jose Luis,

I am planning to release version 2.6.0 in the next few days, for an unrelated reason as I need to add OSGi support.

If you like I can merge the Pull Request as-is tonight or tomorrow (as it's pretty much complete anyway), and then make those two changes myself? Or if you prefer I can release 2.6.0 without this feature, and you can make these changes yourself, and then I can merge it and make a 2.7.0 release soon after. Which would you prefer?

**Comment by jlpedrosa** [comment]
Hi
What ever you preffer, I am happy with both.

**Comment by npgall** [comment]
Okay I will go go for option 1 to merge now, thanks!

**Comment by jlpedrosa** [comment]
Thanks!

**Comment by npgall** [comment]
Done! Version 2.6.0 has been released and it includes these methods. Many thanks Jose Luis for contributing this feature! The new version should sync to Maven Central in a few hours.
